Exit with nonzero if any task failed

GaryShen2008 opened the pull request:
Add an arg '--allow_failure'
Save the query reports into an array
Check the query reports after running and exit with non-zero value to represent a failure of the script if no flag '--allow_failure'
Print out the query which status is 'Failed' or 'CompletedWithFailedTask'

Close #147

GaryShen2008 committed:
Add nofailure arg

Exit 1 if any task failed
CI should use nofailure

Signed-off-by: Gary Shen <gashen@nvidia.com>

GaryShen2008 committed:
fix missing parenthesis

Signed-off-by: Gary Shen <gashen@nvidia.com>

GaryShen2008 committed:
import sys

Signed-off-by: Gary Shen <gashen@nvidia.com>

GaryShen2008 committed:
Fix the queryStatus check

Add more print

Signed-off-by: Gary Shen <gashen@nvidia.com>

GaryShen2008 committed:
fix comment

Signed-off-by: Gary Shen <gashen@nvidia.com>

gerashegalov commented on nds/PysparkBenchReport.py:
check/update copyright year in modified files

GaryShen2008 replied:
Updated copyright year.

gerashegalov commented on nds/nds_power.py:
```diff
@@ -303,6 +307,19 @@ def run_query_stream(input_prefix,
+                    print("====== Queries with failure ======")
+                print("{} status: {}".format(q.summary['query'], q.summary['queryStatus']))
+                exit_code = 1
+        if exit_code == 1:
```
nit to future-proof against other exit_code values: `if not exit_code:`

GaryShen2008 replied:
Thanks, improved the check code.

GaryShen2008 committed:
Update copyright year

Check exit_code not zero

Signed-off-by: Gary Shen <gashen@nvidia.com>

gerashegalov reviewed (dismissed):
LGTM

jlowe commented on nds/nds_power.py:
```diff
@@ -303,6 +307,19 @@ def run_query_stream(input_prefix,
         time_df = spark_session.createDataFrame(data=execution_time_list, schema = header)
         time_df.coalesce(1).write.csv(extra_time_log_output_path)
 
+    if not nofailure:
```
This option name is very confusing.  To me, "nofailure" means we don't want any query failures, but this option only takes effect when it is _not_ set which seems backwards.  Suggest renaming this parameter to better reflect its semantics.

jlowe replied:
Ah, I see it's documented below.  Still confusing to me, but I'm not going to block it.  Something like `allow_failures` is a more appropriate name, IMHO.

GaryShen2008 replied:
updated to use "allow_failure".

jlowe commented on nds/nds_power.py:
```diff
@@ -303,6 +307,19 @@ def run_query_stream(input_prefix,
+        for q in queries_reports:
+            if not q.is_success():
+                if exit_code == 0:
+                    print("====== Queries with failure ======")
```
We only want to report a summary of failing queries sometimes?  Wondering why this isn't always happening, regardless of the option setting.  The option documentation implies all it does is force a zero exit code, but it does more than that.

GaryShen2008 replied:
Yeah, sounds odd.
I updated to always print out the failure information regardless of the option.
The allow_failure parameter only decides to exit non-zero to fail the script.

GaryShen2008 committed:
Rename allow_failure

Always print out the failure
Just exit non-zero code without allow_failure parameter

Signed-off-by: Gary Shen <gashen@nvidia.com>

GaryShen2008 committed:
Fix the condition

Signed-off-by: Gary Shen <gashen@nvidia.com>

jlowe commented on nds/nds_power.py:
```diff
@@ -364,13 +382,15 @@ def load_properties(filename):
                         '"_part1" and "_part2" suffix for the following query names: ' +
                         'query14, query23, query24, query39. e.g. query14_part1, query39_part2')
+    parser.add_argument('--allow_failure',
+                        action='store_true',
```
We want to avoid reporting failed queries via the exit code by default?  This is the opposite of what I would expect.

GaryShen2008 replied:
No, the default is without using "--allow_failure" parameter.
So the failed queries or tasks cause the exit code non-zero.
Only the CI jobs need to use this parameter to skip the failure then report all results.

GaryShen2008 replied:
Hi @jlowe, is it ok to merge or does it require any change?
